File: android-agent/app/src/main/java/com/androidscript/agent/runtime/Value.kt
package com.androidscript.agent.runtime

import kotlin.math.abs

/**
 * Value system for AndroidScript runtime
 * Represents all possible value types in the language
 */

// Type aliases for clarity
typealias NativeFunction = (List<Value>) -> Value
typealias ValueArray = MutableList<Value>
typealias ValueMap = MutableMap<String, Value>

// ========================================================================
// Value Type Enum
// ========================================================================

enum class ValueType {
    NIL,
    BOOLEAN,
    INTEGER,
    FLOAT,
    STRING,
    ARRAY,
    OBJECT,
    FUNCTION,
    NATIVE_FUNCTION,
    DEVICE
}

// ========================================================================
// Device Reference (for multi-device support)
// ========================================================================

data class DeviceRef(
    val serial: String = "",
    val model: String = "",
    var screenWidth: Int = 0,
    var screenHeight: Int = 0,
    val androidVersion: String = "",
    var nativeHandle: Any? = null  // Platform-specific handle
)

// ========================================================================
// Function Object (user-defined functions)
// ========================================================================

data class FunctionObject(
    val parameters: List<String>,
    val body: BlockStmt,
    val closure: Environment?
)

// ========================================================================
// Main Value Class (sealed class hierarchy)
// ========================================================================

sealed class Value {
    abstract val type: ValueType

    // Type checking methods
    fun isNil() = this is NilValue
    fun isBool() = this is BoolValue
    fun isInt() = this is IntValue
    fun isFloat() = this is FloatValue
    fun isNumber() = isInt() || isFloat()
    fun isString() = this is StringValue
    fun isArray() = this is ArrayValue
    fun isObject() = this is ObjectValue
    fun isFunction() = this is FunctionValue
    fun isNativeFunction() = this is NativeFunctionValue
    fun isDevice() = this is DeviceValue
    fun isCallable() = isFunction() || isNativeFunction()

    // Type conversion methods (with runtime checks)
    fun asBool(): Boolean = when (this) {
        is BoolValue -> value
        else -> throw RuntimeException("Value is not a boolean")
    }

    fun asInt(): Long = when (this) {
        is IntValue -> value
        is FloatValue -> value.toLong()
        else -> throw RuntimeException("Value is not a number")
    }

    fun asFloat(): Double = when (this) {
        is FloatValue -> value
        is IntValue -> value.toDouble()
        else -> throw RuntimeException("Value is not a number")
    }

    fun asString(): String = when (this) {
        is StringValue -> value
        else -> throw RuntimeException("Value is not a string")
    }

    fun asArray(): ValueArray = when (this) {
        is ArrayValue -> elements
        else -> throw RuntimeException("Value is not an array")
    }

    fun asObject(): ValueMap = when (this) {
        is ObjectValue -> properties
        else -> throw RuntimeException("Value is not an object")
    }

    fun asDevice(): DeviceRef = when (this) {
        is DeviceValue -> device
        else -> throw RuntimeException("Value is not a device")
    }

    fun asFunction(): FunctionObject = when (this) {
        is FunctionValue -> function
        else -> throw RuntimeException("Value is not a function")
    }

    fun asNativeFunction(): NativeFunction = when (this) {
        is NativeFunctionValue -> function
        else -> throw RuntimeException("Value is not a native function")
    }

    // Truthiness (for conditionals)
    fun isTruthy(): Boolean = when (this) {
        is NilValue -> false
        is BoolValue -> value
        is IntValue -> value != 0L
        is FloatValue -> value != 0.0
        is StringValue -> value.isNotEmpty()
        is ArrayValue -> elements.isNotEmpty()
        is ObjectValue -> properties.isNotEmpty()
        else -> true
    }

    // String representation
    fun typeString(): String = when (this) {
        is NilValue -> "nil"
        is BoolValue -> "boolean"
        is IntValue -> "integer"
        is FloatValue -> "float"
        is StringValue -> "string"
        is ArrayValue -> "array"
        is ObjectValue -> "object"
        is FunctionValue -> "function"
        is NativeFunctionValue -> "native_function"
        is DeviceValue -> "device"
    }

    override fun toString(): String = when (this) {
        is NilValue -> "null"
        is BoolValue -> if (value) "true" else "false"
        is IntValue -> value.toString()
        is FloatValue -> value.toString()
        is StringValue -> value
        is ArrayValue -> "[${elements.joinToString(", ") { it.toString() }}]"
        is ObjectValue -> "{${properties.entries.joinToString(", ") { "${it.key}: ${it.value}" }}}"
        is DeviceValue -> "Device(${device.serial})"
        is FunctionValue -> "<function>"
        is NativeFunctionValue -> "<native function>"
    }

    // ========================================================================
    // Arithmetic Operators
    // ========================================================================

    operator fun plus(other: Value): Value {
        // String concatenation
        if (this is StringValue || other is StringValue) {
            return StringValue(this.toString() + other.toString())
        }

        // Numeric addition
        if (this.isFloat() || other.isFloat()) {
            return FloatValue(this.asFloat() + other.asFloat())
        }
        if (this.isInt() && other.isInt()) {
            return IntValue(this.asInt() + other.asInt())
        }

        throw RuntimeException("Invalid operands for +")
    }

    operator fun minus(other: Value): Value {
        if (this.isFloat() || other.isFloat()) {
            return FloatValue(this.asFloat() - other.asFloat())
        }
        if (this.isInt() && other.isInt()) {
            return IntValue(this.asInt() - other.asInt())
        }
        throw RuntimeException("Invalid operands for -")
    }

    operator fun times(other: Value): Value {
        if (this.isFloat() || other.isFloat()) {
            return FloatValue(this.asFloat() * other.asFloat())
        }
        if (this.isInt() && other.isInt()) {
            return IntValue(this.asInt() * other.asInt())
        }
        throw RuntimeException("Invalid operands for *")
    }

    operator fun div(other: Value): Value {
        if (this.isFloat() || other.isFloat()) {
            val divisor = other.asFloat()
            if (divisor == 0.0) throw RuntimeException("Division by zero")
            return FloatValue(this.asFloat() / divisor)
        }
        if (this.isInt() && other.isInt()) {
            val divisor = other.asInt()
            if (divisor == 0L) throw RuntimeException("Division by zero")
            return IntValue(this.asInt() / divisor)
        }
        throw RuntimeException("Invalid operands for /")
    }

    operator fun rem(other: Value): Value {
        if (!this.isInt() || !other.isInt()) {
            throw RuntimeException("Modulo requires integer operands")
        }
        val divisor = other.asInt()
        if (divisor == 0L) throw RuntimeException("Modulo by zero")
        return IntValue(this.asInt() % divisor)
    }

    // ========================================================================
    // Comparison Operators
    // ========================================================================

    fun equals(other: Value): Boolean {
        if (this.type != other.type) return false

        return when (this) {
            is NilValue -> true
            is BoolValue -> value == (other as BoolValue).value
            is IntValue -> value == (other as IntValue).value
            is FloatValue -> abs(value - (other as FloatValue).value) < 1e-10
            is StringValue -> value == (other as StringValue).value
            is ArrayValue -> elements === (other as ArrayValue).elements  // Reference equality
            is ObjectValue -> properties === (other as ObjectValue).properties
            is DeviceValue -> device.serial == (other as DeviceValue).device.serial
            else -> false
        }
    }

    fun notEquals(other: Value): Boolean = !equals(other)

    fun lessThan(other: Value): Boolean {
        if (this.isNumber() && other.isNumber()) {
            return this.asFloat() < other.asFloat()
        }
        if (this is StringValue && other is StringValue) {
            return value < other.value
        }
        throw RuntimeException("Invalid operands for <")
    }

    fun lessEqual(other: Value): Boolean = lessThan(other) || equals(other)
    fun greaterThan(other: Value): Boolean = !lessEqual(other)
    fun greaterEqual(other: Value): Boolean = !lessThan(other)

    // ========================================================================
    // Unary Operators
    // ========================================================================

    operator fun unaryMinus(): Value {
        return when (this) {
            is IntValue -> IntValue(-value)
            is FloatValue -> FloatValue(-value)
            else -> throw RuntimeException("Invalid operand for unary -")
        }
    }

    operator fun not(): Value = BoolValue(!isTruthy())

    // ========================================================================
    // Array/Object Access
    // ========================================================================

    operator fun get(index: Int): Value {
        if (this !is ArrayValue) throw RuntimeException("Value is not an array")
        if (index < 0 || index >= elements.size) {
            throw RuntimeException("Array index out of bounds: $index")
        }
        return elements[index]
    }

    operator fun get(key: String): Value {
        if (this !is ObjectValue) throw RuntimeException("Value is not an object")
        return properties[key] ?: NilValue
    }

    operator fun set(index: Int, value: Value) {
        if (this !is ArrayValue) throw RuntimeException("Value is not an array")
        if (index < 0 || index >= elements.size) {
            throw RuntimeException("Array index out of bounds: $index")
        }
        elements[index] = value
    }

    operator fun set(key: String, value: Value) {
        if (this !is ObjectValue) throw RuntimeException("Value is not an object")
        properties[key] = value
    }

    // ========================================================================
    // Length Operation
    // ========================================================================

    fun length(): Int = when (this) {
        is ArrayValue -> elements.size
        is StringValue -> value.length
        is ObjectValue -> properties.size
        else -> throw RuntimeException("Value does not have a length")
    }
}

// ========================================================================
// Concrete Value Types
// ========================================================================

object NilValue : Value() {
    override val type = ValueType.NIL
}

data class BoolValue(val value: Boolean) : Value() {
    override val type = ValueType.BOOLEAN
}

data class IntValue(val value: Long) : Value() {
    override val type = ValueType.INTEGER
}

data class FloatValue(val value: Double) : Value() {
    override val type = ValueType.FLOAT
}

data class StringValue(val value: String) : Value() {
    override val type = ValueType.STRING
}

data class ArrayValue(val elements: ValueArray = mutableListOf()) : Value() {
    override val type = ValueType.ARRAY

    fun push(value: Value) = elements.add(value)

    fun pop(): Value {
        if (elements.isEmpty()) throw RuntimeException("Array is empty")
        return elements.removeAt(elements.size - 1)
    }
}

data class ObjectValue(val properties: ValueMap = mutableMapOf()) : Value() {
    override val type = ValueType.OBJECT

    fun hasKey(key: String): Boolean = properties.containsKey(key)

    fun keys(): List<String> = properties.keys.toList()
}

data class FunctionValue(val function: FunctionObject) : Value() {
    override val type = ValueType.FUNCTION
}

data class NativeFunctionValue(val function: NativeFunction) : Value() {
    override val type = ValueType.NATIVE_FUNCTION
}

data class DeviceValue(val device: DeviceRef) : Value() {
    override val type = ValueType.DEVICE
}

// ========================================================================
// Factory Methods (companion object pattern)
// ========================================================================

object Values {
    fun makeNil() = NilValue
    fun makeBool(b: Boolean) = BoolValue(b)
    fun makeInt(i: Long) = IntValue(i)
    fun makeInt(i: Int) = IntValue(i.toLong())
    fun makeFloat(d: Double) = FloatValue(d)
    fun makeString(s: String) = StringValue(s)
    fun makeArray(elements: ValueArray = mutableListOf()) = ArrayValue(elements)
    fun makeObject(properties: ValueMap = mutableMapOf()) = ObjectValue(properties)
    fun makeFunction(function: FunctionObject) = FunctionValue(function)
    fun makeNativeFunction(function: NativeFunction) = NativeFunctionValue(function)
    fun makeDevice(device: DeviceRef) = DeviceValue(device)
}

File: ios-agent/iOSAgent/Runtime/Value.swift
import Foundation

/**
 * Value system for AndroidScript runtime
 * Swift implementation using enums with associated values
 */

// Type aliases
typealias NativeFunction = ([Value]) -> Value
typealias ValueArray = [Value]
typealias ValueMap = [String: Value]

// MARK: - Device Reference

struct DeviceRef {
    var serial: String = ""
    var model: String = ""
    var screenWidth: Int = 0
    var screenHeight: Int = 0
    var androidVersion: String = ""
    var nativeHandle: Any?
}

// MARK: - Function Object

class FunctionObject {
    let parameters: [String]
    let body: BlockStmt
    let closure: Environment?

    init(parameters: [String], body: BlockStmt, closure: Environment?) {
        self.parameters = parameters
        self.body = body
        self.closure = closure
    }
}

// MARK: - Environment (for closures)

class Environment {
    private var values: [String: Value] = [:]
    private let parent: Environment?

    init(parent: Environment? = nil) {
        self.parent = parent
    }

    func define(name: String, value: Value) {
        values[name] = value
    }

    func get(name: String) throws -> Value {
        if let value = values[name] {
            return value
        }

        if let parent = parent {
            return try parent.get(name: name)
        }

        throw RuntimeError.undefinedVariable(name)
    }

    func assign(name: String, value: Value) throws {
        if values[name] != nil {
            values[name] = value
            return
        }

        if let parent = parent {
            try parent.assign(name: name, value: value)
            return
        }

        throw RuntimeError.undefinedVariable(name)
    }

    func exists(name: String) -> Bool {
        if values[name] != nil {
            return true
        }
        return parent?.exists(name: name) ?? false
    }

    func isGlobal() -> Bool {
        return parent == nil
    }

    func clear() {
        values.removeAll()
    }
}

// MARK: - Value Enum

enum Value {
    case `nil`
    case bool(Bool)
    case int(Int64)
    case float(Double)
    case string(String)
    case array(ValueArray)
    case object(ValueMap)
    case function(FunctionObject)
    case nativeFunction(NativeFunction)
    case device(DeviceRef)

    // MARK: - Type Checking

    func isNil() -> Bool {
        if case .nil = self { return true }
        return false
    }

    func isBool() -> Bool {
        if case .bool = self { return true }
        return false
    }

    func isInt() -> Bool {
        if case .int = self { return true }
        return false
    }

    func isFloat() -> Bool {
        if case .float = self { return true }
        return false
    }

    func isNumber() -> Bool {
        return isInt() || isFloat()
    }

    func isString() -> Bool {
        if case .string = self { return true }
        return false
    }

    func isArray() -> Bool {
        if case .array = self { return true }
        return false
    }

    func isObject() -> Bool {
        if case .object = self { return true }
        return false
    }

    func isFunction() -> Bool {
        if case .function = self { return true }
        return false
    }

    func isNativeFunction() -> Bool {
        if case .nativeFunction = self { return true }
        return false
    }

    func isCallable() -> Bool {
        return isFunction() || isNativeFunction()
    }

    // MARK: - Type Conversion

    func asBool() throws -> Bool {
        guard case .bool(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not a boolean")
        }
        return value
    }

    func asInt() throws -> Int64 {
        switch self {
        case .int(let value):
            return value
        case .float(let value):
            return Int64(value)
        default:
            throw RuntimeError.typeMismatch("Value is not a number")
        }
    }

    func asFloat() throws -> Double {
        switch self {
        case .float(let value):
            return value
        case .int(let value):
            return Double(value)
        default:
            throw RuntimeError.typeMismatch("Value is not a number")
        }
    }

    func asString() throws -> String {
        guard case .string(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not a string")
        }
        return value
    }

    func asArray() throws -> ValueArray {
        guard case .array(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not an array")
        }
        return value
    }

    func asObject() throws -> ValueMap {
        guard case .object(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not an object")
        }
        return value
    }

    func asFunction() throws -> FunctionObject {
        guard case .function(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not a function")
        }
        return value
    }

    func asNativeFunction() throws -> NativeFunction {
        guard case .nativeFunction(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not a native function")
        }
        return value
    }

    func asDevice() throws -> DeviceRef {
        guard case .device(let value) = self else {
            throw RuntimeError.typeMismatch("Value is not a device")
        }
        return value
    }

    // MARK: - Truthiness

    func isTruthy() -> Bool {
        switch self {
        case .nil:
            return false
        case .bool(let value):
            return value
        case .int(let value):
            return value != 0
        case .float(let value):
            return value != 0.0
        case .string(let value):
            return !value.isEmpty
        case .array(let value):
            return !value.isEmpty
        case .object(let value):
            return !value.isEmpty
        default:
            return true
        }
    }

    // MARK: - String Representation

    func toString() -> String {
        switch self {
        case .nil:
            return "null"
        case .bool(let value):
            return value ? "true" : "false"
        case .int(let value):
            return String(value)
        case .float(let value):
            return String(value)
        case .string(let value):
            return value
        case .array(let elements):
            let items = elements.map { $0.toString() }.joined(separator: ", ")
            return "[\(items)]"
        case .object(let map):
            let items = map.map { "\($0.key): \($0.value.toString())" }.joined(separator: ", ")
            return "{\(items)}"
        case .device(let device):
            return "Device(\(device.serial))"
        case .function:
            return "<function>"
        case .nativeFunction:
            return "<native function>"
        }
    }

    func typeString() -> String {
        switch self {
        case .nil: return "nil"
        case .bool: return "boolean"
        case .int: return "integer"
        case .float: return "float"
        case .string: return "string"
        case .array: return "array"
        case .object: return "object"
        case .function: return "function"
        case .nativeFunction: return "native_function"
        case .device: return "device"
        }
    }

    // MARK: - Arithmetic Operators

    static func +(lhs: Value, rhs: Value) throws -> Value {
        // String concatenation
        if lhs.isString() || rhs.isString() {
            return .string(lhs.toString() + rhs.toString())
        }

        // Numeric addition
        if lhs.isFloat() || rhs.isFloat() {
            return .float(try lhs.asFloat() + try rhs.asFloat())
        }

        if lhs.isInt() && rhs.isInt() {
            return .int(try lhs.asInt() + try rhs.asInt())
        }

        throw RuntimeError.invalidOperands("Invalid operands for +")
    }

    static func -(lhs: Value, rhs: Value) throws -> Value {
        if lhs.isFloat() || rhs.isFloat() {
            return .float(try lhs.asFloat() - try rhs.asFloat())
        }

        if lhs.isInt() && rhs.isInt() {
            return .int(try lhs.asInt() - try rhs.asInt())
        }

        throw RuntimeError.invalidOperands("Invalid operands for -")
    }

    static func *(lhs: Value, rhs: Value) throws -> Value {
        if lhs.isFloat() || rhs.isFloat() {
            return .float(try lhs.asFloat() * try rhs.asFloat())
        }

        if lhs.isInt() && rhs.isInt() {
            return .int(try lhs.asInt() * try rhs.asInt())
        }

        throw RuntimeError.invalidOperands("Invalid operands for *")
    }

    static func /(lhs: Value, rhs: Value) throws -> Value {
        if lhs.isFloat() || rhs.isFloat() {
            let divisor = try rhs.asFloat()
            if divisor == 0.0 {
                throw RuntimeError.divisionByZero
            }
            return .float(try lhs.asFloat() / divisor)
        }

        if lhs.isInt() && rhs.isInt() {
            let divisor = try rhs.asInt()
            if divisor == 0 {
                throw RuntimeError.divisionByZero
            }
            return .int(try lhs.asInt() / divisor)
        }

        throw RuntimeError.invalidOperands("Invalid operands for /")
    }

    static func %(lhs: Value, rhs: Value) throws -> Value {
        guard lhs.isInt() && rhs.isInt() else {
            throw RuntimeError.invalidOperands("Modulo requires integer operands")
        }

        let divisor = try rhs.asInt()
        if divisor == 0 {
            throw RuntimeError.divisionByZero
        }

        return .int(try lhs.asInt() % divisor)
    }

    // MARK: - Unary Operators

    static prefix func -(value: Value) throws -> Value {
        if value.isInt() {
            return .int(-(try value.asInt()))
        }
        if value.isFloat() {
            return .float(-(try value.asFloat()))
        }
        throw RuntimeError.invalidOperands("Invalid operand for unary -")
    }

    static prefix func !(value: Value) -> Value {
        return .bool(!value.isTruthy())
    }

    // MARK: - Comparison Operators

    func equals(_ other: Value) -> Bool {
        switch (self, other) {
        case (.nil, .nil):
            return true
        case (.bool(let a), .bool(let b)):
            return a == b
        case (.int(let a), .int(let b)):
            return a == b
        case (.float(let a), .float(let b)):
            return abs(a - b) < 1e-10
        case (.string(let a), .string(let b)):
            return a == b
        default:
            return false
        }
    }

    func lessThan(_ other: Value) throws -> Bool {
        if self.isNumber() && other.isNumber() {
            return try self.asFloat() < other.asFloat()
        }
        if self.isString() && other.isString() {
            return try self.asString() < other.asString()
        }
        throw RuntimeError.invalidOperands("Invalid operands for <")
    }

    func lessEqual(_ other: Value) throws -> Bool {
        return try lessThan(other) || equals(other)
    }

    func greaterThan(_ other: Value) throws -> Bool {
        return try !lessEqual(other)
    }

    func greaterEqual(_ other: Value) throws -> Bool {
        return try !lessThan(other)
    }

    // MARK: - Array/Object Access

    func length() throws -> Int {
        switch self {
        case .array(let elements):
            return elements.count
        case .string(let value):
            return value.count
        case .object(let map):
            return map.count
        default:
            throw RuntimeError.typeMismatch("Value does not have a length")
        }
    }
}

// MARK: - Runtime Errors

enum RuntimeError: Error {
    case undefinedVariable(String)
    case typeMismatch(String)
    case invalidOperands(String)
    case divisionByZero
    case arrayIndexOutOfBounds(Int)
    case objectKeyNotFound(String)
    case custom(String)

    var localizedDescription: String {
        switch self {
        case .undefinedVariable(let name):
            return "Undefined variable: \(name)"
        case .typeMismatch(let message):
            return message
        case .invalidOperands(let message):
            return message
        case .divisionByZero:
            return "Division by zero"
        case .arrayIndexOutOfBounds(let index):
            return "Array index out of bounds: \(index)"
        case .objectKeyNotFound(let key):
            return "Object key not found: \(key)"
        case .custom(let message):
            return message
        }
    }
}

// MARK: - Control Flow Exceptions

enum ControlFlow: Error {
    case `return`(Value)
    case `break`
    case `continue`
}
